test/puppetlabs/http/client/async_plaintext_test.clj

```diff
@@ -352,4 +356,110 @@
+           (is (= 200 status))
+           (let [instream body
+                 buf (make-array Byte/TYPE 4)
+                 _ (.read instream buf)]
+             ;; Make sure we can read a few chars off of the stream
+             (is (= "xxxx" (String. buf "UTF-8")))
+             ;; Indicate we read some chars
+             (if initial-bytes-read? (deliver initial-bytes-read? true))
+             ;; Read the rest and validate the content
+             (let [final-string (str "xxxx" (slurp instream))]
+               (is (= (str data "xxxx") final-string))))))))))
+
+(defn- clojure-streaming-socket-timeout
+  [data-size opts]
+  (testing (str "clojure streaming socket timeout: " data-size " bytes with opts " opts)
+    (let [data (apply str (repeat (- data-size 4) "x"))]
+      (try
+        (testwebserver/with-test-webserver-and-config
+         (build-content-handler data nil true) port {:shutdown-timeout-seconds 1}
+         (with-open [client (async/create-client {:connect-timeout-milliseconds 100 :socket-timeout-milliseconds 200})]
+           (let [response @(common/get
+                            client
+                            (str "http://localhost:" port "/hello")
+                            opts)
+                 {:keys [body error]} response]
+             (if (= :unbuffered-stream (:as opts))
+               (do
+                 (if error
+                   ;; If there's an error this should behave the same as :stream
+                   (is (instance? SocketTimeoutException error))
+                   ;; If there is no error, let's consume the body to get the exception
+                   (is (thrown? SocketTimeoutException (slurp body)))))
+               (do
+                 (is error)
+                 (is (instance? SocketTimeoutException error)))))))
+        (catch TimeoutException e
+          ;; Expected whenever a server-side failure is generated
+          )))))
+
+(defn- clojure-streaming-connection-error
+  [opts]
+  (testing (str "clojure streaming connect error")
+    (with-open [client (async/create-client {:connect-timeout-milliseconds 100})]
+      (let [response @(common/get
+                       client
+                       (str "http://localhost:" 12345 "/hello")
+                       opts)
+            {:keys [error]} response]
+        (is error)
+        (is (instance? ConnectException error))))))
+
+(deftest clojure-streaming
+  (testing "clojure streaming is consistent with different payload sizes and opts"
+    (testlogging/with-test-logging
+     (doseq [data-size [32 (* 1024) (* 1024 1024)]
+             decompress-body? [false true]
+             as [:unbuffered-stream :stream]]
+       (clojure-streaming-success data-size {:as as :decompress-body decompress-body?})
+       (clojure-streaming-socket-timeout data-size {:as as :decompress-body decompress-body?})
+       (clojure-streaming-connection-error {:as as :decompress-body decompress-body?})))))
```

Comment:
I gave it a rework to try to clarify things. There are really two possible options: the response is totally buffered before it is returned to the caller or it isn't. I've created two separate functions that test these possibilities without an conditionals. Each possible combinations of client options is then pulled into a separate test but passed through one of those two variants.

My hope is that this is a bit less convoluted that before but still allows us to verify that different combinations of options behave in the same way.